### PR TITLE
feat: Add reusable_sync_branch.yml

### PR DESCRIPTION
**Pull request by hiepmai-babylonchain**
This PR introduces a new GitHub Actions workflow named Sync branch. 
The purpose of this workflow is to create a PR for synchronization of a target branch with a base branch. 
This workflow can be invoked from other workflows to create a PR for synchronization of a target branch with a base branch. 
Example usage of the workflow can be found [here](https://github.com/babylonchain/simple-staking/pull/381), in the simple-staking repo

Additionally, this PR adds a CHANGELOG.md file to track all the changes made in the project.

### CODE REVIEW

File: .github/workflows/reusable_sync_branch.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        with:
+          fetch-depth: 0
+      
+      - name: Check for differences between branches
+        id: check_diff
+        run: |
+          base_branch=${{ inputs.base_branch }}
+          target_branch=${{ inputs.target_branch }}
+
+          if git diff --quiet origin/$base_branch..origin/$target_branch; then
```
Comment (jrwbabylonlab) [inline .github/workflows/reusable_sync_branch.yml]: hmm... i'm not fully understanding the git command. 
the dev branch should always be ahead of main branch. in the case of below, will this git diff command still work?
`main` have commits : A, B, C
`dev` have commits: A,B,C,D,E,F

We shall not execute the pipeline to raise PR for the above use case.
But we shall do for below:
`main` have commits : A, B, C
`dev` have commits: C,D,E,F

The commit A & B is the thing we try to merge back into `dev`

I initially thought we shall be using `git log main --not dev`

Comment (hiepmai-babylonchain) [reply .github/workflows/reusable_sync_branch.yml]: Actually, the `git log main --not dev` or `git log $base_branch --not $target_branch` is indeed better in this case.
Will test in my local repo and let you know the result

**Review by jrwbabylonlab — APPROVED** [review]

File: .github/workflows/reusable_sync_branch.yml (hunk 1 above)
Comment (hiepmai-babylonchain) [reply .github/workflows/reusable_sync_branch.yml]: https://github.com/hiepmai-babylonchain/simple-staking/actions/runs/10088262685/job/27893743118
Tested with this run. Can find the log showing all `Unique commits in main that are not in dev`

File: .github/workflows/reusable_sync_branch.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        type: string
+      reviewers:
+        description: 'Comma-separated list of reviewers'
+        required: false
+        default: ''
+        type: string
+
+jobs:
+  create_pr:
+    runs-on: ubuntu-latest
```
Comment (filippos47) [inline .github/workflows/reusable_sync_branch.yml]: We can pin to 22.04 (latest GH ubuntu ver)

File: .github/workflows/reusable_sync_branch.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+          target_branch="${{ inputs.target_branch }}"
+          reviewers="${{ inputs.reviewers }}"
+          repo="${{ github.repository }}"
+
+          pr_exists=$(gh pr list --base "$target_branch" --head "$base_branch" --repo "$repo" --json number --jq '.[0].number')
+
+          if [ -n "$pr_exists" ]; then
+            echo "A pull request already exists: #$pr_exists"
+          else
+            gh pr create --base "$target_branch" \
```
Comment (filippos47) [inline .github/workflows/reusable_sync_branch.yml:63]: Would be nice to include a changelog in the PR description, containing all the commits.
Maybe we can re-use `git log origin/$base_branch --not origin/$target_branch --oneline`?

File: CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# CHANGELOG
```
Comment (filippos47) [inline CHANGELOG.md:1]: Nice job on adding a changelog!

**Review by filippos47 — APPROVED** [review]
Awesome!

File: .github/workflows/reusable_sync_branch.yml (hunk 3 above)
Comment (hiepmai-babylonchain) [reply .github/workflows/reusable_sync_branch.yml:63]: Do you mean replace `pr_exists` for `git log origin/$base_branch --not origin/$target_branch --oneline`? 
The logic is a bit different though. 

- `git log origin/$base_branch --not origin/$target_branch --oneline` is used to only open the PR if there are commits from `main` are not included in `dev`
- `pr_exists` is used to not open again (e.g We already have the PR to merge from `main` to `dev` so we don't need to create another PR)

Comment (filippos47) [reply .github/workflows/reusable_sync_branch.yml:63]: Synced offline :pray:

**Pull request merged**